File: Solutions/189_Rotate_Array/189_Rotate_Array.cs
public class Solution {
    public void Rotate(int[] nums, int k) {
        if(k == 0)
            return;
        
        k %= nums.Length;
        int cutOff = nums.Length - k - 1;
        var res = new List<int>();
        
        for(int i = cutOff + 1; i < nums.Length; i++)
            res.Add(nums[i]);
        
        for(int i = 0; i <= cutOff; i++)
            res.Add(nums[i]);
        
        for(int i = 0; i < nums.Length; i++)
            nums[i] = res[i];
    }
}

//Third fastest
public class Solution {
    public void Rotate(int[] nums, int k) {
        int[] aux = new int[nums.Length];
        int newValueK = k >= nums.Length ? k % nums.Length : k;
        for (int i = 0; i < nums.Length; i++)
        {
            int originIndex = i - newValueK;
            if (originIndex < 0)
            {
                originIndex = nums.Length + originIndex;
            }
            
            aux[i] = nums[originIndex];
        }
        
        for (int i = 0; i < nums.Length; i++)
        {
            nums[i] = aux[i];
        }
    }
}

//Second fastest
public class Solution {
    public void Rotate(int[] nums, int k) {
        int newValueK = k >= nums.Length ? k % nums.Length : k;
        int counter = 0;
        
        for (int i = 0; counter < nums.Length; i++) 
        {        
            int current = i;     
            int aux = nums[i]; // 1
            
            do //1.  1,2,3,4
            {
                int newIndex = i + newValueK; // 4
                if (newIndex >= nums.Length)
                {
                    newIndex = newIndex - nums.Length; // 0
                }

                var oldValue = nums[newIndex]; // 1
                nums[newIndex] = aux; // 3,2,1,4
                aux = oldValue; // 3
                i = newIndex; // 2
                counter++;
            } while (i != current);
        }
    }
}

//Fastest
public class Solution 
{
    public void Rotate(int[] nums, int k) 
    {
        k %= nums.Length;
        Reverse(nums, 0, nums.Length - 1);
        Reverse(nums, 0, k - 1);
        Reverse(nums, k, nums.Length - 1);
    }
    
    private void Reverse(int[] nums, int start, int end)
    {
        while (start < end)
        {
            var temp = nums[start];
            nums[start] = nums[end];
            nums[end] = temp;
            
            start++;
            end--;
        }
    }
}

File: Solutions/189_Rotate_Array/189_Rotate_Array.py
class Solution:
    def rotate(self, nums: List[int], k: int) -> None:
        def inverse(i, j):
            while i < j:
                nums[i], nums[j] = nums[j], nums[i]
                i, j = i + 1, j - 1
        
        n = len(nums)
        k = k % n
        inverse(0, n-1)
        inverse(0, k-1)
        inverse(k, n-1)
        return nums

#Fourth fastest
class Solution:
    def rotate(self, nums: List[int], k: int) -> None:
        """
        Do not return anything, modify nums in-place instead.
        """
        
        ret = [nums[i] for i in range(len(nums))] 
        
        for i in range(len(nums)):
            nums[(i + k)% len(nums)] = ret[i]
                
        
        
#         k = k % len(nums)
        
#         if not k:
#             return nums
        
#         visited = 0 
#         ptr = 0
#         temp = None
        
#         while visited < len(nums):
#             if temp == None:
#                 temp = nums[ptr]
#             else:
#                 temp2 = nums[ptr]
#                 nums[ptr] = temp
#                 temp = temp2
            
#             ptr = (ptr + k) % len(nums)
#             visited += 1

#Third fastest
class Solution:
    def rotate(self, nums: List[int], k: int) -> None:
        """
        Do not return anything, modify nums in-place instead.
        """
        if k == 0:
            return
        
        # if k is greater than the length of the list we don't need to 
        #rotate through the whole thing to get the desired result
        if k > len(nums):
            k = k % len(nums)
            
        prepend = nums[-k:]
        nums[-k:], nums[0:0] = [], prepend


#Second fastest
class Solution:
    def rotate(self, nums: List[int], k: int) -> None:
        n = len(nums)
        k = k % n
        nums[:] = nums[n - k:] + nums[:n - k]

#Fastest
class Solution:
    def rotate(self, nums: List[int], k: int) -> None:
        a = nums[len(nums)-k:]
        a += nums[0: len(nums) - k]
        nums[:] = a